pyoomph/c_output.py

```diff
@@ -487,102 +494,78 @@ def __init__(
+            type_id = the_type.name
+            need_include = False
+        else:
+            defining_file_pair = self.session.get_file_pair_for_type(the_type)
+            type_id = defining_file_pair.id
+            need_include = defining_file_pair is not self
 
-    # May evaluate c_expression several times
-    def emit_incref(self, c_expression: str, the_type: Type) -> str:
-        if isinstance(the_type, UnionType):
-            return f"incref_{self.get_type_c_name(the_type)}({c_expression})"
+        result = f"struct class_{type_id}"
         if (
-            the_type.generic_origin is not None
-            and the_type.generic_origin.generic is OPTIONAL
-        ):
-            value_incref = self.emit_incref(
-                f"({c_expression}).value", the_type.generic_origin.arg
+            the_type.refcounted
+            and not isinstance(the_type, UnionType)
+            and not (
+                the_type.generic_origin is not None
+                and the_type.generic_origin.generic is OPTIONAL
             )
-            return f"do{{ if(!({c_expression}).isnull) {value_incref}; }} while(0)"
-        if the_type.refcounted:
-            return f"incref({c_expression})"
-        return "(void)0"
-
-    def emit_decref(self, c_expression: str, the_type: Type) -> str:
-        if isinstance(the_type, UnionType):
-            return f"decref_{self.get_type_c_name(the_type)}({c_expression})"
-        if (
-            the_type.generic_origin is not None
-            and the_type.generic_origin.generic is OPTIONAL
         ):
-            value_decref = self.emit_decref(
-                f"({c_expression}).value", the_type.generic_origin.arg
-            )
-            return f"(({c_expression}).isnull ? (void)0 : {value_decref})"
-        if the_type.refcounted:
-            return f"decref(({c_expression}), dtor_{self.get_type_c_name(the_type)})"
-        return "(void)0"
-
-    def emit_string(self, value: str) -> str:
-        if value not in self.strings:
-            self.strings[value] = (
-                f"string{len(self.strings)}_" + re.sub(r"[^A-Za-z0-9]", "", value)[:30]
-            )
-
-            # String constants consist of int64_t refcount set to -1,
-            # followed by utf8, followed by zero byte
-            # TODO: is this cross-platform enough?
-            struct_bytes = b"\xff" * 8 + value.encode("utf-8") + b"\0"
+            result += "*"
+        else:
+            can_fwd_declare_in_header = False
 
-            array_content = ", ".join(r"'\x%02x'" % byte for byte in struct_bytes)
-            self.string_defs += f"""
-            static {self.emit_type(STRING)} {self.strings[value]}
-            = (void*)(unsigned char[]){{ {array_content} }};
-            """
-        return self.strings[value]
+        if need_include:
+            self.c_includes.add(defining_file_pair)
+            if can_fwd_declare_in_header and defining_file_pair not in self.h_includes:
```

Comment:
This check is not necessary, as declaring the same thing several times is not an error:

```suggestion
            if can_fwd_declare_in_header:
```